llama-index-integrations/multi_modal_llms/llama-index-multi-modal-llms-nvidia/llama_index/multi_modal_llms/nvidia/base.py

```diff
@@ -0,0 +1,549 @@
+            endpoint (str): The API endpoint to send the request to.
+            messages (Dict[str, Any]): The request payload.
+
+        Returns:
+            Dict[str, Any]: The API response.
+        """
+        async with aiohttp.ClientSession() as session:
+            async with session.post(
+                endpoint,
+                json={"messages": messages, "stream": stream, **kwargs},
+                headers={**self._get_headers(stream=stream), **extra_headers},
+            ) as response:
+                response.raise_for_status()
+                return await response.json()
+
+
+class NVIDIAMultiModal(MultiModalLLM):
+    model: str = Field(description="The Multi-Modal model to use from NVIDIA.")
+    temperature: float = Field(description="The temperature to use for sampling.")
+    max_tokens: Optional[int] = Field(
+        description=" The maximum numbers of tokens to generate, ignoring the number of tokens in the prompt.",
+        gt=0,
+    )
+    timeout: float = Field(
+        default=60.0,
+        description="The timeout, in seconds, for API requests.",
+        ge=0,
+    )
+    api_key: str = Field(default=None, description="The NVIDIA API key.", exclude=True)
+    base_url: str = Field(default=BASE_URL, description="The base URL for NVIDIA API.")
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additional kwargs for the NVIDIA API."
+    )
+    default_headers: Optional[Dict[str, str]] = Field(
+        default=None, description="The default headers for API requests."
+    )
+
+    def __init__(
+        self,
+        model: str = "microsoft/phi-3-vision-128k-instruct",
+        temperature: float = DEFAULT_TEMPERATURE,
+        max_tokens: Optional[int] = 300,
+        nvidia_api_key: Optional[str] = None,
+        api_key: Optional[str] = None,
+        base_url: Optional[str] = BASE_URL,
+        callback_manager: Optional[CallbackManager] = None,
+        default_headers: Optional[Dict[str, str]] = None,
+        **kwargs: Any,
+    ) -> None:
+        api_key = get_from_param_or_env(
+            "api_key",
+            nvidia_api_key or api_key,
+            "NVIDIA_API_KEY",
+            "NO_API_KEY_PROVIDED",
+        )
+
+        is_hosted = base_url in KNOWN_URLS
+
+        if is_hosted and api_key == "NO_API_KEY_PROVIDED":
+            raise "An API key is required for the hosted NIM. This will become an error in 0.2.0."
+
+        super().__init__(
+            model=model,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            api_key=api_key,
+            api_base=base_url,
+            callback_manager=callback_manager,
+            default_headers=default_headers,
+            **kwargs,
+        )
+
+    @property
+    def _client(self) -> NVIDIAClient:
+        _client = NVIDIAClient(**self._get_credential_kwargs())
+        _client._custom_headers = {"User-Agent": "llama-index-multi-modal-llms-nvidia"}
```

Comment:
it appears `_custom_headers` is not used. please move the user agent into NVIDIAClient's _get_headers.